contracts/external/proxies/LiquidatorProxyV1ForSoloMargin.sol

```diff
@@ -0,0 +1,433 @@
+
+        // roll-back the old value
+        uint256 oldValue = cache.owedWei.value.mul(cache.owedPrice);
+        if (cache.owedWei.sign) {
+            cache.supplyValue = cache.supplyValue.sub(oldValue);
+        } else {
+            cache.borrowValue = cache.borrowValue.sub(oldValue);
+        }
+
+        // add the new value
+        uint256 newValue = newOwedWei.value.mul(cache.owedPrice);
+        if (newOwedWei.sign) {
+            cache.supplyValue = cache.supplyValue.add(newValue);
+        } else {
+            cache.borrowValue = cache.borrowValue.add(newValue);
+        }
+
+        cache.owedWei = newOwedWei;
+    }
+
+    /**
+     * Changes heldWei, supplyValue, and borrowValue.
+     * Does not change owedWei and liquidated.
+     */
+    function setHeldWeiTo(
+        Types.Wei memory newHeldWei,
+        Cache memory cache
+    )
+        private
+        pure
+    {
+        // roll-back the old value
+        uint256 oldValue = cache.heldWei.value.mul(cache.heldPrice);
+        if (cache.heldWei.sign) {
+            cache.supplyValue = cache.supplyValue.sub(oldValue);
+        } else {
+            cache.borrowValue = cache.borrowValue.sub(oldValue);
+        }
+
+        // add the new value
+        uint256 newValue = newHeldWei.value.mul(cache.heldPrice);
+        if (newHeldWei.sign) {
+            cache.supplyValue = cache.supplyValue.add(newValue);
+        } else {
+            cache.borrowValue = cache.borrowValue.add(newValue);
+        }
+
+        cache.heldWei = newHeldWei;
+    }
+
+    function liquidateUntilNegOwedAndPosHeld(
+        Cache memory cache
+    )
+        private
+        pure
+    {
+        bool negOwed = !cache.owedWei.isPositive();
+        bool posHeld = !cache.heldWei.isNegative();
+
+        if (negOwed && posHeld) {
+            return;
+        }
+
+        // true if it takes longer for the liquidator owed balance to become negative than it takes
+        // the liquidator held balance to become positive.
+        bool owedGoesToZeroLast;
+        if (negOwed) {
+            owedGoesToZeroLast = false;
+        } else if (posHeld) {
+            owedGoesToZeroLast = true;
+        } else {
+            // owed is still positive and held is still negative
+            owedGoesToZeroLast =
+                cache.owedWei.value.mul(cache.owedPriceAdj) >
+                cache.heldWei.value.mul(cache.heldPrice);
+        }
+
+        if (owedGoesToZeroLast) {
+            Types.Wei memory heldWeiDelta = owedWeiToHeldWei(cache.owedWei, cache);
+            setHeldWeiTo(cache.heldWei.add(heldWeiDelta), cache);
+            setOwedWeiTo(Types.zeroWei(), cache);
+        } else {
+            Types.Wei memory owedWeiDelta = heldWeiToOwedWei(cache.heldWei, cache);
+            setOwedWeiTo(cache.owedWei.add(owedWeiDelta), cache);
+            setHeldWeiTo(Types.zeroWei(), cache);
+        }
+    }
+
+    function figureOutWhereToStop(
```

Comment:
Can we make a more official name for this function haha?

Maybe `calculateLiquidationAmount` or something?